File: src/main/kotlin/customfit/ai/kotlinclient/network/HttpClient.kt
package customfit.ai.kotlinclient.network

import customfit.ai.kotlinclient.constants.CFConstants
import customfit.ai.kotlinclient.config.core.CFConfig
import customfit.ai.kotlinclient.core.error.CFResult
import customfit.ai.kotlinclient.core.error.ErrorHandler
import customfit.ai.kotlinclient.logging.Timber
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.atomic.AtomicInteger
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject

class HttpClient(private val cfConfig: CFConfig? = null) {
    companion object {
        private const val SOURCE = "HttpClient"
    }
    
    // Use atomics to allow thread-safe updates
    private val connectionTimeout = AtomicInteger(cfConfig?.networkConnectionTimeoutMs ?: 
                                                 CFConstants.Network.CONNECTION_TIMEOUT_MS)
    private val readTimeout = AtomicInteger(cfConfig?.networkReadTimeoutMs ?: 
                                           CFConstants.Network.READ_TIMEOUT_MS)

    /**
     * Updates the connection timeout setting
     *
     * @param timeoutMs new timeout in milliseconds
     */
    fun updateConnectionTimeout(timeoutMs: Int) {
        require(timeoutMs > 0) { "Timeout must be greater than 0" }
        connectionTimeout.set(timeoutMs)
        Timber.d("Updated connection timeout to $timeoutMs ms")
    }

    /**
     * Updates the read timeout setting
     *
     * @param timeoutMs new timeout in milliseconds
     */
    fun updateReadTimeout(timeoutMs: Int) {
        require(timeoutMs > 0) { "Timeout must be greater than 0" }
        readTimeout.set(timeoutMs)
        Timber.d("Updated read timeout to $timeoutMs ms")
    }

    /**
     * Performs an HTTP request with robust error handling
     */
    suspend fun <T> performRequest(
            url: String,
            method: String,
            headers: Map<String, String> = emptyMap(),
            body: String? = null,
            responseHandler: suspend (HttpURLConnection) -> T?
    ): T? =
            withContext(Dispatchers.IO) {
                var connection: HttpURLConnection? = null
                try {
                    Timber.d("API CALL: $method request to $url")
                    connection = URL(url).openConnection() as HttpURLConnection
                    connection.requestMethod = method
                    connection.connectTimeout = connectionTimeout.get()
                    connection.readTimeout = readTimeout.get()
                    connection.doInput = true
                    connection.instanceFollowRedirects = true
                    connection.setRequestProperty("User-Agent", "CustomFit-SDK/1.0 Kotlin")

                    headers.forEach { (key, value) -> 
                        connection.setRequestProperty(key, value)
                    }

                    if (body != null) {
                        connection.doOutput = true
                        connection.outputStream.use { it.write(body.toByteArray()) }
                    }

                    val response = responseHandler(connection)
                    connection.disconnect()
                    return@withContext response
                } catch (e: Exception) {
                    // Use our robust error handling system
                    Timber.e("API ERROR: ${e.javaClass.simpleName} - ${e.message}")
                    val category = ErrorHandler.handleException(
                        e, 
                        "Error making $method request to $url", 
                        SOURCE,
                        ErrorHandler.ErrorSeverity.HIGH
                    )
                    
                    connection?.disconnect()
                    null
                }
            }

    /**
     * Fetches metadata from a URL with improved error handling
     */
    suspend fun fetchMetadata(url: String): CFResult<Map<String, String>> =
            performRequest(url, "GET") { conn ->
                Timber.d("EXECUTING GET METADATA REQUEST")
                if (conn.responseCode == HttpURLConnection.HTTP_OK) {
                    val metadata = conn.headerFields.let { headers ->
                        mapOf(
                                CFConstants.Http.HEADER_LAST_MODIFIED to (headers["Last-Modified"]?.firstOrNull() ?: ""),
                                CFConstants.Http.HEADER_ETAG to (headers["ETag"]?.firstOrNull() ?: "")
                        )
                    }
                    // Since we're using GET instead of HEAD, we need to read the response to ensure the connection is released
                    conn.inputStream.bufferedReader().readText()
                    Timber.d("GET METADATA SUCCESSFUL: $metadata")
                    CFResult.success(metadata)
                } else {
                    val message = "Failed to fetch metadata from $url: ${conn.responseCode}"
                    Timber.w("GET METADATA FAILED: $message")
                    ErrorHandler.handleError(
                        message, 
                        SOURCE, 
                        ErrorHandler.ErrorCategory.NETWORK
                    )
                    CFResult.error(message, code = conn.responseCode, category = ErrorHandler.ErrorCategory.NETWORK)
                }
            } ?: CFResult.error("Network error fetching metadata from $url", category = ErrorHandler.ErrorCategory.NETWORK)

    /**
     * Fetches JSON from a URL with improved error handling
     */
    suspend fun fetchJson(url: String): CFResult<JsonObject> =
            performRequest(url, "GET") { conn ->
                Timber.d("EXECUTING GET JSON REQUEST")
                if (conn.responseCode == HttpURLConnection.HTTP_OK) {
                    try {
                        val responseText = conn.inputStream.bufferedReader().readText()
                        // Parse using kotlinx.serialization and cast to JsonObject
                        val jsonElement = Json.parseToJsonElement(responseText)
                        if (jsonElement is JsonObject) {
                            Timber.d("GET JSON SUCCESSFUL")
                            CFResult.success(jsonElement)
                        } else {
                            val message = "Parsed JSON from $url is not an object"
                            Timber.w("GET JSON FAILED: $message")
                            ErrorHandler.handleError(
                                message, 
                                SOURCE, 
                                ErrorHandler.ErrorCategory.SERIALIZATION
                            )
                            CFResult.error(message, category = ErrorHandler.ErrorCategory.SERIALIZATION)
                        }
                    } catch (e: Exception) {
                        Timber.e("GET JSON FAILED: ${e.message}")
                        ErrorHandler.handleException(
                            e, 
                            "Error parsing JSON response from $url", 
                            SOURCE,
                            ErrorHandler.ErrorSeverity.HIGH
                        )
                        CFResult.error("Error parsing JSON response", e, category = ErrorHandler.ErrorCategory.SERIALIZATION)
                    }
                } else {
                    val message = "Failed to fetch JSON from $url: ${conn.responseCode}"
                    Timber.w("GET JSON FAILED: $message")
                    ErrorHandler.handleError(
                        message, 
                        SOURCE, 
                        ErrorHandler.ErrorCategory.NETWORK
                    )
                    CFResult.error(message, code = conn.responseCode, category = ErrorHandler.ErrorCategory.NETWORK)
                }
            } ?: CFResult.error("Network error fetching JSON from $url", category = ErrorHandler.ErrorCategory.NETWORK)

    /**
     * Posts JSON to a URL with improved error handling and detailed logging
     */
    suspend fun postJson(url: String, payload: String): CFResult<Boolean> =
            performRequest(url, "POST", mapOf(CFConstants.Http.HEADER_CONTENT_TYPE to CFConstants.Http.CONTENT_TYPE_JSON), payload) { conn ->
                Timber.d("EXECUTING POST JSON REQUEST")
                val responseCode = conn.responseCode

                try {
                    if (responseCode == HttpURLConnection.HTTP_OK ||
                                    responseCode == HttpURLConnection.HTTP_ACCEPTED
                    ) {
                        val responseBody = conn.inputStream.bufferedReader().readText()
                        Timber.d("POST JSON SUCCESSFUL")
                        CFResult.success(true)
                    } else {
                        val errorBody = conn.errorStream?.bufferedReader()?.readText() ?: "No error body"
                        
                        // Use our error handling system
                        val message = "API error response: ${conn.responseCode}"
                        Timber.w("POST JSON FAILED: $message - $errorBody")
                        ErrorHandler.handleError(
                            "$message - $errorBody", 
                            SOURCE, 
                            ErrorHandler.ErrorCategory.NETWORK, 
                            ErrorHandler.ErrorSeverity.HIGH
                        )
                        
                        Timber.e("Error: $errorBody")
                        CFResult.error(message, code = conn.responseCode, category = ErrorHandler.ErrorCategory.NETWORK)
                    }
                } catch (e: Exception) {
                    Timber.e("POST JSON FAILED: ${e.message}")
                    ErrorHandler.handleException(
                        e, 
                        "Failed to read API response", 
                        SOURCE,
                        ErrorHandler.ErrorSeverity.HIGH
                    )
                    CFResult.error("Failed to read API response", e, category = ErrorHandler.ErrorCategory.NETWORK)
                }
            } ?: CFResult.error("Network error posting JSON to $url", category = ErrorHandler.ErrorCategory.NETWORK)

    /**
     * Performs a HEAD request to efficiently check for metadata changes
     * Only fetches headers without downloading the full response body
     */
    suspend fun makeHeadRequest(url: String): CFResult<Map<String, String>> {
        return withContext(Dispatchers.IO) {
            try {
                Timber.i("API POLL: HEAD request to $url")
                val connection = URL(url).openConnection() as HttpURLConnection
                connection.requestMethod = "HEAD"
                connection.connectTimeout = CFConstants.Network.CONNECTION_TIMEOUT_MS
                connection.readTimeout = CFConstants.Network.READ_TIMEOUT_MS
                Timber.d("EXECUTING HEAD REQUEST")
                
                // Apply retry and timeout logic if needed
                val responseCode = connection.responseCode
                
                if (responseCode == HttpURLConnection.HTTP_OK) {
                    // Extract headers
                    val headers = mutableMapOf<String, String>()
                    
                    // Get Last-Modified header which is crucial for caching
                    connection.getHeaderField(CFConstants.Http.HEADER_LAST_MODIFIED)?.let {
                        headers[CFConstants.Http.HEADER_LAST_MODIFIED] = it
                    }
                    
                    // Get ETag header for additional caching support
                    connection.getHeaderField(CFConstants.Http.HEADER_ETAG)?.let {
                        headers[CFConstants.Http.HEADER_ETAG] = it
                    }
                    
                    Timber.i("API POLL: HEAD request successful - Last-Modified: ${headers[CFConstants.Http.HEADER_LAST_MODIFIED]}, ETag: ${headers[CFConstants.Http.HEADER_ETAG]}")
                    return@withContext CFResult.success(headers)
                } else {
                    Timber.w("API POLL: HEAD request failed with code: $responseCode")
                    return@withContext CFResult.error(
                        "HEAD request failed with code: $responseCode",
                        category = ErrorHandler.ErrorCategory.NETWORK
                    )
                }
            } catch (e: Exception) {
                Timber.e("API POLL: HEAD request exception: ${e.message}")
                return@withContext CFResult.error(
                    "HEAD request failed with exception: ${e.message}",
                    e,
                    category = ErrorHandler.ErrorCategory.NETWORK
                )
            }
        }
    }

}

File: customfit-flutter-client-sdk/lib/src/network/http_client.dart
import 'dart:async';

import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';

import '../config/core/cf_config.dart';
import '../core/error/cf_result.dart';
import '../core/error/error_category.dart';
import '../core/error/error_handler.dart';
import '../core/error/error_severity.dart';
import '../constants/cf_constants.dart';
import '../core/logging/logger.dart';

/// HTTP client implementation mirroring Kotlin's HttpClient
class HttpClient {
  // ignore: unused_field
  final CFConfig _config;
  late final Dio _dio;
  int _connectionTimeoutMs;
  int _readTimeoutMs;

  static const String _SOURCE = 'HttpClient';

  HttpClient(this._config)
      : _connectionTimeoutMs = _config.networkConnectionTimeoutMs,
        _readTimeoutMs = _config.networkReadTimeoutMs {
    _initializeDio();
  }

  void _initializeDio() {
    _dio = Dio(BaseOptions(
      connectTimeout: Duration(milliseconds: _connectionTimeoutMs),
      receiveTimeout: Duration(milliseconds: _readTimeoutMs),
      headers: {
        CFConstants.http.headerContentType: CFConstants.http.contentTypeJson
      },
    ));

    // Add custom interceptor for full request/response/error logging
    _dio.interceptors.add(
      InterceptorsWrapper(
        onRequest: (options, handler) {
          Logger.d('==== CF API REQUEST ====');
          Logger.d('URL: ${options.uri}');
          Logger.d('Method: ${options.method}');
          Logger.d('Headers: ${options.headers}');
          Logger.d('Payload: ${options.data}');
          Logger.d('=======================');
          return handler.next(options);
        },
        onResponse: (response, handler) {
          Logger.d('==== CF API RESPONSE ====');
          Logger.d('Status: ${response.statusCode}');
          Logger.d('Data: ${response.data}');
          Logger.d('========================');
          return handler.next(response);
        },
        onError: (DioError e, handler) {
          Logger.e('==== CF API ERROR ====');
          Logger.e('URL: ${e.requestOptions.uri}');
          Logger.e('Error: ${e.error}');
          Logger.e('Response: ${e.response}');
          Logger.e('======================');
          return handler.next(e);
        },
      ),
    );

    // Add logging interceptor
    _dio.interceptors.add(LogInterceptor(
      request: true,
      requestHeader: true,
      requestBody: true,
      responseHeader: true,
      responseBody: true,
      error: true,
      logPrint: (log) => Logger.d('HTTP: $log'),
    ));

    Logger.i(
        'HttpClient initialized with connectTimeout=$_connectionTimeoutMs ms, readTimeout=$_readTimeoutMs ms');
  }

  /// Update connection timeout
  void updateConnectionTimeout(int timeoutMs) {
    if (timeoutMs <= 0) throw ArgumentError('Timeout must be > 0');
    _connectionTimeoutMs = timeoutMs;
    _dio.options.connectTimeout = Duration(milliseconds: timeoutMs);
    Logger.i('Updated connection timeout to $timeoutMs ms');
  }

  /// Update read timeout
  void updateReadTimeout(int timeoutMs) {
    if (timeoutMs <= 0) throw ArgumentError('Timeout must be > 0');
    _readTimeoutMs = timeoutMs;
    _dio.options.receiveTimeout = Duration(milliseconds: timeoutMs);
    Logger.i('Updated read timeout to $timeoutMs ms');
  }

  /// GET request for metadata (Last-Modified, ETag)
  Future<CFResult<Map<String, String>>> fetchMetadata(String url,
      {String? lastModified, String? etag}) async {
    try {
      Logger.i('GET $url for metadata');

      // Add caching headers to avoid unnecessary network calls
      final headers = {
        'Content-Type': 'application/json',
        'Cache-Control': 'no-cache',
      };

      // Add conditional request headers if available
      if (lastModified != null &&
          lastModified.isNotEmpty &&
          lastModified != 'unchanged') {
        headers['If-Modified-Since'] = lastModified;
        Logger.d('Using If-Modified-Since: $lastModified');
      }

      if (etag != null && etag.isNotEmpty && etag != 'unchanged') {
        headers['If-None-Match'] = etag;
        Logger.d('Using If-None-Match: $etag');
      }

      final options = Options(
        headers: headers,
        // Set validateStatus to accept 304 Not Modified responses
        validateStatus: (status) =>
            status != null && (status >= 200 && status < 300 || status == 304),
      );

      Logger.d('🔎 API HTTP: GET metadata for $url');
      final resp = await _dio.get(url, options: options);

      // Handle 304 Not Modified (return the same headers)
      if (resp.statusCode == 304) {
        Logger.i('🔎 API HTTP: Metadata unchanged (304 Not Modified)');
        return CFResult.success({
          CFConstants.http.headerLastModified: lastModified ?? 'unchanged',
          CFConstants.http.headerEtag: etag ?? 'unchanged',
        });
      }

      // Handle 200 OK with headers
      if (resp.statusCode == 200) {
        final headers = resp.headers;
        final resultHeaders = {
          CFConstants.http.headerLastModified:
              headers.value('Last-Modified') ?? '',
          CFConstants.http.headerEtag: headers.value('ETag') ?? '',
        };
        Logger.i('🔎 API HTTP: Got metadata headers: $resultHeaders');
        return CFResult.success(resultHeaders);
      } else {
        final msg = 'Failed GET metadata $url: ${resp.statusCode}';
        Logger.w('🔎 API HTTP: $msg');
        ErrorHandler.handleError(msg,
            source: _SOURCE,
            category: ErrorCategory.network,
            severity: ErrorSeverity.medium);
        return CFResult.error(msg,
            code: resp.statusCode ?? 0, category: ErrorCategory.network);
      }
    } catch (e) {
      final errorMsg = 'Error GET metadata $url: ${e.toString()}';
      Logger.e('🔎 API HTTP: $errorMsg');
      ErrorHandler.handleException(e, errorMsg,
          source: _SOURCE, severity: ErrorSeverity.high);
      return CFResult.error(errorMsg,
          exception: e, category: ErrorCategory.network);
    }
  }

  /// GET request for a JSON object
  Future<CFResult<Map<String, dynamic>>> fetchJson(String url) async {
    try {
      Logger.i('GET $url');
      final resp = await _dio.get(url);
      if (resp.statusCode == 200) {
        final data = resp.data;
        if (data is Map<String, dynamic>) {
          Logger.i('🔎 API HTTP: Successfully fetched JSON from $url');
          return CFResult.success(data);
        } else {
          final msg = 'GET $url JSON not object';
          Logger.w('🔎 API HTTP: $msg - Response data is not a JSON object');
          ErrorHandler.handleError(msg,
              source: _SOURCE,
              category: ErrorCategory.serialization,
              severity: ErrorSeverity.medium);
          return CFResult.error(msg, category: ErrorCategory.serialization);
        }
      } else {
        final msg = 'Failed GET $url: ${resp.statusCode}';
        Logger.w('🔎 API HTTP: $msg');
        ErrorHandler.handleError(msg,
            source: _SOURCE,
            category: ErrorCategory.network,
            severity: ErrorSeverity.high);
        return CFResult.error(msg,
            code: resp.statusCode ?? 0, category: ErrorCategory.network);
      }
    } catch (e) {
      Logger.e('🔎 API HTTP: Error GET $url: ${e.toString()}');
      ErrorHandler.handleException(e, 'Error GET $url',
          source: _SOURCE, severity: ErrorSeverity.high);
      return CFResult.error('Network error GET $url: ${e.toString()}',
          exception: e, category: ErrorCategory.network);
    }
  }

  /// POST raw JSON string
  Future<CFResult<bool>> postJson(String url, String payload) async {
    const sep = '===== API RESPONSE =====';
    try {
      Logger.i('POST $url');

      // Log payload size
      Logger.i('POST request payload size: ${payload.length} bytes');

      // Log based on endpoint type
      if (url.contains("summary")) {
        Logger.i('📊 SUMMARY HTTP: POST request');
      } else if (url.contains("cfe") || url.contains("events")) {
        Logger.i('🔔 TRACK HTTP: POST request to event API');
        Logger.i('🔔 TRACK HTTP: Request body size: ${payload.length} bytes');
      }

      final resp = await _dio.post(
        url,
        data: payload,
        options: Options(headers: {
          CFConstants.http.headerContentType: CFConstants.http.contentTypeJson
        }),
      );

      Logger.d(sep);
      Logger.d('Status: ${resp.statusCode}');

      if (resp.statusCode == 200 || resp.statusCode == 202) {
        // Log based on endpoint type
        if (url.contains("summary")) {
          Logger.i('📊 SUMMARY HTTP: Response code: ${resp.statusCode}');
          Logger.i('📊 SUMMARY HTTP: Summary successfully sent to server');
        } else if (url.contains("cfe") || url.contains("events")) {
          Logger.i('🔔 TRACK HTTP: Response code: ${resp.statusCode}');
          Logger.i('🔔 TRACK HTTP: Events successfully sent to server');
        }

        Logger.d(resp.data.toString());
        return CFResult.success(true);
      } else {
        final body = resp.data?.toString() ?? 'No error body';
        final msg = 'Error POST $url: ${resp.statusCode}';

        // Log based on endpoint type
        if (url.contains("summary")) {
          Logger.w('📊 SUMMARY HTTP: Error code: ${resp.statusCode}');
          Logger.w('📊 SUMMARY HTTP: Error body: $body');
        } else if (url.contains("cfe") || url.contains("events")) {
          Logger.w('🔔 TRACK HTTP: Error code: ${resp.statusCode}');
          Logger.w('🔔 TRACK HTTP: Error body: $body');
        }

        ErrorHandler.handleError('$msg – $body',
            source: _SOURCE,
            category: ErrorCategory.network,
            severity: ErrorSeverity.high);
        Logger.e('Error: $body');
        return CFResult.error(msg,
            code: resp.statusCode ?? 0, category: ErrorCategory.network);
      }
    } catch (e) {
      // Log based on endpoint type
      if (url.contains("summary")) {
        Logger.e('📊 SUMMARY HTTP: Exception: ${e.toString()}');
      } else if (url.contains("cfe") || url.contains("events")) {
        Logger.e('🔔 TRACK HTTP: Exception: ${e.toString()}');
      }

      ErrorHandler.handleException(e, 'Error POST $url',
          source: _SOURCE, severity: ErrorSeverity.high);
      return CFResult.error('Network error POST $url: ${e.toString()}',
          exception: e, category: ErrorCategory.network);
    } finally {
      Logger.d(sep);
    }
  }

  /// Generic POST with dynamic body & query
  Future<CFResult<dynamic>> post(
    String url, {
    dynamic data,
    Map<String, dynamic>? queryParameters,
    Options? options,
  }) async {
    const sep = '===== API RESPONSE =====';
    try {
      Logger.i('POST $url');

      // Determine if this is a tracking or summary request
      bool isTracking = url.contains("events") || url.contains("cfe");
      bool isSummary = url.contains("summary");

      if (isTracking) {
        Logger.i('🔔 TRACK HTTP: POST request to: $url');
        if (data != null) {
          Logger.i(
              '🔔 TRACK HTTP: Request body size: ${data.toString().length} bytes');
        }
      } else if (isSummary) {
        Logger.i('📊 SUMMARY HTTP: POST request to: $url');
      }

      final resp = await _dio.post(
        url,
        data: data,
        queryParameters: queryParameters,
        options: options,
      );

      Logger.d(sep);
      Logger.d('Status: ${resp.statusCode}');

      if (resp.statusCode == 200 || resp.statusCode == 202) {
        if (isTracking) {
          Logger.i('🔔 TRACK HTTP: Response code: ${resp.statusCode}');
          Logger.i('🔔 TRACK HTTP: Events successfully sent to server');
        } else if (isSummary) {
          Logger.i('📊 SUMMARY HTTP: Response code: ${resp.statusCode}');
          Logger.i('📊 SUMMARY HTTP: Summary successfully sent to server');
        }

        return CFResult.success(resp.data);
      } else {
        final body = resp.data?.toString() ?? 'No error body';
        final msg = 'Error POST $url: ${resp.statusCode}';

        if (isTracking) {
          Logger.w('🔔 TRACK HTTP: Error code: ${resp.statusCode}');
          Logger.w('🔔 TRACK HTTP: Error body: $body');
        } else if (isSummary) {
          Logger.w('📊 SUMMARY HTTP: Error code: ${resp.statusCode}');
          Logger.w('📊 SUMMARY HTTP: Error body: $body');
        }

        ErrorHandler.handleError('$msg – $body',
            source: _SOURCE,
            category: ErrorCategory.network,
            severity: ErrorSeverity.high);
        Logger.e('Error: $body');
        return CFResult.error(msg,
            code: resp.statusCode ?? 0, category: ErrorCategory.network);
      }
    } catch (e) {
      bool isTracking = url.contains("events") || url.contains("cfe");
      bool isSummary = url.contains("summary");

      if (isTracking) {
        Logger.e('🔔 TRACK HTTP: Exception: ${e.toString()}');
      } else if (isSummary) {
        Logger.e('📊 SUMMARY HTTP: Exception: ${e.toString()}');
      }

      ErrorHandler.handleException(e, 'Error POST $url',
          source: _SOURCE, severity: ErrorSeverity.high);
      return CFResult.error('Network error POST $url: ${e.toString()}',
          exception: e, category: ErrorCategory.network);
    } finally {
      Logger.d(sep);
    }
  }

  /// HEAD wrapper returning raw Dio Response
  Future<CFResult<Response>> head(String url, {Options? options}) async {
    try {
      debugPrint('HEAD $url');
      final resp = await _dio.head(url, options: options);
      return CFResult.success(resp);
    } catch (e) {
      ErrorHandler.handleException(e, 'Error HEAD $url',
          source: _SOURCE, severity: ErrorSeverity.high);
      return CFResult.error('Network error HEAD $url: ${e.toString()}',
          exception: e, category: ErrorCategory.network);
    }
  }
}
